tensorflow/python/eager/def_function.py

```diff
@@ -1168,6 +1170,39 @@ def f_sig(y):
     concrete._garbage_collector.release()  # pylint: disable=protected-access
     return concrete
 
+  def validate_annotations(self):
+    """Ensures type annotations and input_signature match.
+
+    Raises:
+      ValueError when type annotations do not match input_signature.
+    """
+    if self.input_signature is None:
+      return
+    for tensor in self.input_signature:
+      if tensor.__class__ != tensor_spec.TensorSpec:
+        return
+    input_signature_dtypes = tuple(tensor.dtype for tensor
+                                   in self.input_signature)
+    if len(self.function_spec.arg_names) != len(input_signature_dtypes):
+      return
+
+    arg_dtype_list = zip(self.function_spec.arg_names,
+                         input_signature_dtypes)
+    # A dict mapping from arg name to input signature dtype
+    arg_to_dtype = dict(arg_dtype_list)
+    type_annotations = self.function_spec.fullargspec.annotations
+    for arg, annot in type_annotations.items():
+      if (not hasattr(typing, "_GenericAlias") or
```

Comment:
Let's avoid reliance on private symbols. I think we can use `hasattr(type(annot), '__args__')`

Reply:
Sounds good, we can check for `if not hasattr(annot, '__args__'):`